jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java

```diff
@@ -591,69 +592,69 @@ public void doHandle(String target, Request baseRequest, HttpServletRequest requ
             return _servletPathMap.getMatch(target);
         }
 
-        if (_servletNameMap == null)
-            return null;
         ServletHolder holder = _servletNameMap.get(target);
         if (holder == null)
             return null;
         return new MappedResource<>(null, holder);
     }
 
+    private FilterChain linkFilterChain(ServletHolder servletHolder, FilterHolder filterHolder, FilterChain chain)
+    {
+        if (chain == null)
+            chain = new ServletFilterChain(servletHolder);
+        FilterChain next = newFilterChain(filterHolder.getFilter(), chain);
+        return filterHolder.isAsyncSupported() ? next : new NotAsyncFilterChain(filterHolder, chain);
+    }
+
+    /**
+     * Create a FilterChain that calls the passed filter with the passed chain
+     * @param filter The filter to invoke
+     * @param chain The chain to pass to the filter
+     * @return A FilterChain that invokes the filter with the chain
+     */
+    protected FilterChain newFilterChain(Filter filter, FilterChain chain)
+    {
+        return new Chain(filter, chain);
+    }
+
     protected FilterChain getFilterChain(Request baseRequest, String pathInContext, ServletHolder servletHolder)
     {
         String key = pathInContext == null ? servletHolder.getName() : pathInContext;
         int dispatch = FilterMapping.dispatch(baseRequest.getDispatcherType());
 
-        if (_filterChainsCached && _chainCache != null)
+        if (_filterChainsCached)
         {
             FilterChain chain = _chainCache[dispatch].get(key);
             if (chain != null)
                 return chain;
         }
 
-        // Build list of filters (list of FilterHolder objects)
-        List<FilterHolder> filters = new ArrayList<>();
-
-        // Path filters
-        if (pathInContext != null && _filterPathMappings != null)
-        {
-            for (FilterMapping filterPathMapping : _filterPathMappings)
-            {
-                if (filterPathMapping.appliesTo(pathInContext, dispatch))
-                    filters.add(filterPathMapping.getFilterHolder());
-            }
-        }
+        FilterChain chain = null;
 
-        // Servlet name filters
         if (servletHolder != null && _filterNameMappings != null && !_filterNameMappings.isEmpty())
```

Comment:
The Servlet Spec requires that the first filters in the chain are the url style mappings, and then the servlet name mappings. This does it the other way around, doesn't it?

Reply:
I would hope we fail some of our unit tests then.
We should have basic tests that validate these kinds of requirements from the servlet spec.
Unless Greg has done some kind of trickery and his "Reverse mappings" statement is really "mappings done in reverse" making the change he made valid again.

Reply:
@janbartel The lists are in reverse so we apply the inner onion skin first. The order the filters are executed is not changed.